README.md
# Nitro Sniper Enhanced [![CodeFactor](https://www.codefactor.io/repository/github/giorgiobrux/nitro-sniper-enhanced/badge)](https://www.codefactor.io/repository/github/giorgiobrux/nitro-sniper-enhanced) ![GitHub issues](https://img.shields.io/github/issues/giorgiobrux/nitro-sniper-enhanced) [![GPLv3 license](https://img.shields.io/badge/License-GPLv3-blue.svg)](http://perso.crans.org/besson/LICENSE.html)
<img alt="img" align="right" src="https://user-images.githubusercontent.com/18328525/92536909-19ea5000-f23b-11ea-8fb7-524b4ba22f26.png">
Snipes nitro gift codes; with alt-support.<br>
A fork of the (previously-MIT licensed) sniper by slow.
  
# Features
<img alt="img" align="right" src="https://user-images.githubusercontent.com/18328525/93030855-88a61f80-f626-11ea-882f-3136ca00d26e.png">

- Beautiful colors to see quickly what you need to see.
- Multi-token support with all codes redeemed on your main.
- Notifications from the mobile discord app work if you don't use your main account for sniping too.
- Removes non-alphanumeric chars automatically from codes and tries to redeem.
- Supports `code&,junk` -> `code` obfuscation.
- Auto-detects fake codes by length or probability check.
- It remembers all codes it tries and won't check them again. (cached on disk).
- Optional webhook support; Avatar and username set by default.
- Will snipe privnotes searching for codes and saving the content to a file.
# Disclaimer
The use of selfbots is against the [discord TOS](https://discord.com/terms) and [guidelines](https://discord.com/guidelines).  
Before proceeding, make sure you understand the implications and can accept the consequences.

# Installation methods
#### Heroku (recommended)
The good guys at heroku have apparently blacklisted this repo.<br>
In order to deploy an heroku instance:
- Fork the repo.
- Go to https://dashboard.heroku.com/new?template=https://github.com/YOURUSERNAME/nitro-sniper-enhanced making sure to change `YOURUSERNAME` in the url to your actual github username.
- To update your fork, you can use the `Fetch upstream` button on github.

> ℹ️ While directly using someone else's fork is possible, it is heavily discouraged for novice users because malicious code can be added and if one becomes very popular it will also be blacklisted by heroku.

After deployment, make sure you go to your app -> Configure dynos/Resources and turn off `web` and start `worker`.<br>
You can then see logs by clicking on `More` (top right) and then `View logs`.<br>
To update see [here](https://github.com/GiorgioBrux/nitro-sniper-enhanced/issues/7#issuecomment-692116471) or just remove your app and recreate it.
#### Repl.it
Go to your dashboard, click on the plus button, import from github and type `giorgiobrux/nitro-sniper-enhanced`. <br>
Edit the .env file, making sure to set `replit` to `true`. <br>
No one else will be able to see the .env file even with public repos, so your tokens are safe. <br>
**IMPORTANT**: You need to use a service that periodically pings your repl.it app like [UptimeRobot](https://uptimerobot.com/), or else it will go in sleep mode. <br>
To update, run `git pull` from the shell.
#### Local
- Make sure [Node](https://nodejs.org/en/) 14+ is installed on your system and open a command prompt/terminal.
- Run `git clone https://github.com/giorgiobrux/nitro-sniper-enhanced nitro-sniper`
- Run `cd nitro-sniper`
- Run `npm install`
- Edit the .env file.
- Run `node .` <br>
To update, run `git pull`  
#### Docker
[Available on DockerHub](https://hub.docker.com/r/giorgiobrux/nitro-sniper) <br>
To update, use [watchtower](https://github.com/containrrr/watchtower) or similar. 

# Contributing
Pull requests are welcome.  
If you contribute, all your code will be licensed under GPL-3.0-or-later.  

# Tips
- Try to get a low latency to discord servers as there can be competition with other snipers.
- This is technically a self-bot: mentioning this in a discord chat is enough to make your account reportable to Trust & Safety.
- Running more than one instance or different snipers is an easy way to get your account deactivated.
- If you get a lot of nitro in a short amount of time, turn the sniper off for a while.
- Before using, consider the moral implications of stealing gifts from communities you have nothing to do with.

# License
[GPL-3.0-or-later](https://www.gnu.org/licenses/)  
Modified work Copyright (C) 2020-2021 All nitro-sniper-enhanced contributors  
Original work Copyright (C) 2020 slow | Sublicensed according to the MIT license available [here](https://opensource.org/licenses/MIT) or in the [LICENSE.md](https://github.com/GiorgioBrux/nitro-sniper-enhanced/blob/master/LICENSE.md) file.



CHANGELOG.md
## [1.11.5](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.11.4...v1.11.5) (2021-08-30)


### Bug Fixes

* bad package-lock.json ([5b24365](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/5b243653c1cba2e3865555db27a8ef6251428b32)), closes [#141](https://github.com/giorgiobrux/nitro-sniper-enhanced/issues/141)

## [1.11.4](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.11.3...v1.11.4) (2021-08-25)


### Bug Fixes

* don't crash if guildTokens is undefined ([9b5c009](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/9b5c00917780ee91cc43dacea0ce51270a802008)), closes [#137](https://github.com/giorgiobrux/nitro-sniper-enhanced/issues/137)

## [1.11.3](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.11.2...v1.11.3) (2021-05-15)


### Bug Fixes

* fix broken chalk ([04ed990](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/04ed990ae1a1ac56a98c2871038376f3fc67d9c9))

## [1.11.2](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.11.1...v1.11.2) (2021-05-15)


### Bug Fixes

* **deps:** update dependency dotenv to v9 ([0782199](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/0782199d4eee1a75044a9da5fadcce412333a3d2))

## [1.11.1](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.11.0...v1.11.1) (2021-04-02)


### Bug Fixes

* fix npm i ([9241bdf](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/9241bdf620dd601bffffaa96145ba6aaf48c1b53))

# [1.11.0](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.10.0...v1.11.0) (2021-04-02)


### Bug Fixes

* fix error when invalid url are provided ([494120c](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/494120ca1ae4cc591f04414cb0818aa5720383d2))


### Features

* temp.pm support ([b58d907](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/b58d907e31f5c6b562fa72667df413db4317f512))

# [1.10.0](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.9.3...v1.10.0) (2021-04-02)


### Bug Fixes

* fix broken dependency ([45e362d](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/45e362d46174ba7fa897e08baca1dcc99be50524))


### Features

* add mainStatus ([6209cd7](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/6209cd78b7e2f4db03193f8fb415343c5bcd01c4)), closes [#33](https://github.com/giorgiobrux/nitro-sniper-enhanced/issues/33)
* ignore messages with irrealistic number of gifts ([df9d11f](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/df9d11fe46d8a8ec3be0062efef00f3038a16b76))

## [1.9.3](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.9.2...v1.9.3) (2021-03-16)


### Bug Fixes

* fix spam of connection messages ([257edd6](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/257edd6f87a20810da91842971f782c2242504d3)), closes [#57](https://github.com/giorgiobrux/nitro-sniper-enhanced/issues/57)

## [1.9.2](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.9.1...v1.9.2) (2021-03-08)


### Bug Fixes

* fix broken chalk ([e6480af](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/e6480aff49dac2ff6d1a22721660338b81330bc1))

## [1.9.1](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.9.0...v1.9.1) (2021-02-23)


### Bug Fixes

* fix wrong console.log ([60e9252](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/60e92526a7292265cd3e46ff796dcc6813c8fbc8))

# [1.9.0](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.8.8...v1.9.0) (2021-02-21)


### Features

* automatic get latest useragent ([cb01c58](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/cb01c58ba922c951b6e17ed5733bf467d48af54e))

## [1.8.8](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.8.7...v1.8.8) (2021-02-21)


### Bug Fixes

* fix end is not defined ([1e10db7](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/1e10db7eef81406db782c539d591fedeaebe24f6))

## [1.8.7](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.8.6...v1.8.7) (2021-02-21)


### Bug Fixes

* revert back to v6 ([fe4b33b](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/fe4b33bf90029bef1eefe766463b94e113110344))

## [1.8.6](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.8.5...v1.8.6) (2021-02-21)


### Bug Fixes

* fix invalid bitfield? ([2bd5788](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/2bd5788044be6d0f8acc102cda31b026f4f7e00d))

## [1.8.5](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.8.4...v1.8.5) (2021-02-21)


### Bug Fixes

* fix double webhook_userid warning ([66211ed](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/66211edd1f1bdfaad0073ba2bfc8711d3ddcce9e))
* ping only in the nitro webhook ([90edc99](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/90edc99a9569347b96e57a09d68fae3f5ea189f7))

## [1.8.4](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.8.3...v1.8.4) (2021-02-21)


### Bug Fixes

* **replit:** fix wrong comparison ([b9c8234](https://github.com/giorgiobrux/nitro-sniper-enhanced/commit/b9c823461c0b65d6c0d8b3a712962c75da244939))

## [1.8.3](https://github.com/giorgiobrux/nitro-sniper-enhanced/compare/v1.8.2...v1.8.3) (2021-02-21)


### Bug Fixes


src/webhooks.js
const { WebhookClient, MessageEmbed } = require("discord.js-light");
const logging = require('./logging/logging');

class Webhook {
  constructor(webhookUrl, type) {
    this.a = 2;
    if (webhookUrl === "") {
      logging.info(`{blueBright The ${type} webhook is empty, skipping...}`);
      return null;
    }

    const webhooktoken = /[^/]*$/.exec(webhookUrl)[0];
    const webhookid = webhookUrl.replace(
      /^.*\/(?=[^/]*\/[^/]*$)|\/[^/]*$/g,
      ""
    );
    if (
      webhooktoken == null ||
      webhookid == null ||
      webhooktoken.length < webhookid.length ||
      !/https:\/\/(ptb\.|canary\.|)(discordapp|discord)\.com\/api\/webhooks\/[0-9]+\/.+/g.test(
        webhookUrl
      )
    ) {
      logging.error(`{red The ${type} webhook url is not valid. Skipping...}`);
      return null;
    }
    logging.info(`{blueBright Using ${type} webhook with id: [${webhookid}] and token: [${webhooktoken}].}`);
    this.client = new WebhookClient(webhookid, webhooktoken);
    this.a = 1;
  }

  show_a() {
    console.log(`a = ${this.client}`);
  }

  send_webhook_nitro(
    res_type,
    guild,
    giver,
    tokenname,
    timetaken,
    code,
    msgurl,
    webhookping_userid
  ) {
    if (!this.client) return;
    const embed = new MessageEmbed()
      .setTitle(`Sniped gift successfully!`)
      .setColor("#1ce829")
      .addField("Where", `${guild}`, true)
      .addField("Account used", `${tokenname}`, true)
      .addField("Giver", `${giver}`, true)
      .addField("Time taken", `${timetaken}`, true)
      .addField("Type of sub", `${res_type}`, true)
      .addField("Giftcode", `${code}`, true)
      .addField("​", `[Click here for the message.](${msgurl})`, false);
    this.client
      .send(`${webhookping_userid ? `<@${webhookping_userid}>` : ""}`, {
        username: "Nitro Sniper",
        avatarURL:
          "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcS0JCyNz1WwaTkXB3jcr0MlMLIwXAsHjhoIRw&usqp=CAU",
        embeds: [embed],
      })
      .catch((err) => {
        logging.error(`{red Tried to send nitro webhook embed but got error: ${err}.}`);
      });
  }

  send_webhook_notes(noteweb, guild, giver, tokenname, content, msgurl) {
    if (!this.client) return;
    const embed = new MessageEmbed()
      .setTitle(`Sniped note successfully!`)
      .setColor("#1ce5e8")
      .addField("Where", `${guild}`, true)
      .addField("Account used", `${tokenname}`, true)
      .addField("​", "​", true) // Dummy field
      .addField("Sender", `${giver}`, true)
      .addField("Type", `${noteweb}`, true)
      .addField("Content", `${content}`, false)
      .addField("​", `[Click here for the message.](${msgurl})`, false);
    this.client
      .send(``, {
        username: "Nitro Sniper",
        avatarURL:
          "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcS0JCyNz1WwaTkXB3jcr0MlMLIwXAsHjhoIRw&usqp=CAU",
        embeds: [embed],
      })
      .catch((err) => {
        logging.error(`{red Tried to send notes webhook embed but got error: ${err}.}`);
      });
  }
}

module.exports = { Webhook };


src/logging/logging.js
const chalk = require("chalk");

function chalkish(parts, ...substitutions) {
  // https://gist.github.com/bennadel/71eb3670ccf51daabf1658fb13b337b1
  const rawResults = [];
  const cookedResults = [];

  const partsLength = parts.length;
  const substitutionsLength = substitutions.length;

  for (let i = 0; i < partsLength; i += 1) {
    rawResults.push(parts.raw[i]);
    cookedResults.push(parts[i]);

    if (i < substitutionsLength) {
      rawResults.push(substitutions[i]);
      cookedResults.push(substitutions[i]);
    }
  }

  const chalkParts = [cookedResults.join("")];
  chalkParts.raw = [rawResults.join("")];

  return chalk(chalkParts);
}

const splash = () => {
  console.log(
    `%c    _   ___ __                _____       _                   
     / | / (_) /__________     / ___/____  (_)___  ___  _____   
    /  |/ / / __/ ___/ __ \\    \\__ \\/ __ \\/ / __ \\/ _ \\/ ___/   
   / /|  / / /_/ /  / /_/ /   ___/ / / / / / /_/ /  __/ /       
  /_/ |_/_/\\__/_/_  \\____/   /____/_/ /_/_/ .___/\\___/_/      __
    ___  ____  / /_  ____ _____  ________/_/___/ /  ___  ____/ /
   / _ \\/ __ \\/ __ \\/ __ \`/ __ \\/ ___/ _ \\/ __  /  / _ \\/ __  / 
  /  __/ / / / / / / /_/ / / / / /__/  __/ /_/ /  /  __/ /_/ /  
  \\___/_/ /_/_/ /_/\\__,_/_/ /_/\\___/\\___/\\__,_/   \\___/\\__,_/ `,
    "font-family:monospace"
  );
};

const genericPrefixed = (msg) => {
  console.log(chalkish`{magenta [Nitro Sniper]} ${msg}`);
};

const info = (msg) => {
  genericPrefixed(`{cyan (INFO)} ${msg}`);
};

const error = (msg) => {
  genericPrefixed(`{rgb(242,46,46) (ERROR)} ${msg}`);
};

const warning = (msg) => {
  genericPrefixed(`{yellowBright (WARNING)} ${msg}`);
};

const fatal = (msg) => {
  genericPrefixed(`{rgb(242,46,46).bold (FATAL ERROR)} ${msg}`);
};

const success = (msg) => {
  genericPrefixed(`{rgb(28,232,41) [+]} ${msg}`);
};

module.exports = {
  splash,
  info,
  error,
  warning,
  genericPrefixed,
  fatal,
  success,
};


src/index.js
/*
Nitro Sniper enhanced ed.
Modified work Copyright (C) 2020-2021 Nitro-sniper-enhanced contributors
Original work Copyright (C) 2020 slow | Sublicensed according to the MIT license available at <https://opensource.org/licenses/MIT> or in the LICENSE.md file in the root folder.

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>. */

global.regex = new RegExp(
  /(discord\.gift\/|discord\.com\/gifts\/|discordapp\.com\/gifts\/)[^\s]+/gim
);
const privnote = new RegExp(/(?<=privnote.com\/)[^\s]+/);
const temp_pm = new RegExp(/(?<=temp.pm\/\?)[^\s]+/);

require("dotenv").config();
const axios = require("axios").default;
const rateLimit = require("axios-rate-limit");

const http = rateLimit(axios.create(), {
  maxRequests: 10,
  perMilliseconds: 60000,
});

const syncrq = require("sync-request");
const fs = require("fs");

const { Client } = require("discord.js-light");
const notes = require("./notes/all");
const { Webhook } = require("./webhooks");
const { version } = require("../package.json");
const logging = require("./logging/logging");

const tokens = process.env.guildTokens?.split(",").filter((item) => item);

let { permanentCache } = process.env;

const {
  useMain,
  mainToken,
  nitro_webhookUrl,
  notes_webhookUrl,
  legitimacyCheck,
  obfuscationCheck,
  notesCheck,
  writeNotes,
  webhookping_userid,
  replit,
  tokenStatus,
  mainStatus,
} = process.env;

let usedTokens = [];

if (useMain === "true" && mainToken != null) tokens.unshift(mainToken);
logging.splash();

logging.info("{blueBright Welcome!}");
logging.info(
  `{blueBright Running version} {blueBright.bold ${version}}{blueBright .}`
);
logging.info(
  "{redBright This program is licensed under GPL-3.0-or-later and provided free of charge at https://github.com/GiorgioBrux/nitro-sniper-enhanced.}"
);

global.userAgent = (() =>
  axios({
    url: "https://www.whatismybrowser.com/guides/the-latest-user-agent/chrome",
    method: "GET",
  })
    .then((res) => {
      const r = res.data
        .match(/Mozilla\/5\.0 \(Windows NT 10\.0; Win64; x64\)[^<]*/g)
        .toString();
      logging.info(`{blueBright Setting latest chrome user agent: ${r}}`);
      return r;
    })
    .catch((err) => {
      logging.error(
        `{red There was an error fetching the latest chrome user agent: ${err
          .toString()
          .substring(7, err.length)}. Using default...}`
      );
      return "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36";
    }))();

if (
  !webhookping_userid ||
  webhookping_userid.length === 0 ||
  webhookping_userid === ""
) {
  logging.warning(
    "{rgb(255,245,107) webhookping_userid is not set correctly or is undefined. Defaulting to none.}"
  );
} else {
  logging.info(
    `{blueBright The userid [${webhookping_userid}] will be pinged in the nitro_webhook.}`
  );
}
const nitro_webhookclient = new Webhook(nitro_webhookUrl, "nitro");
const notes_webhookclient = new Webhook(notes_webhookUrl, "notes");

if (!tokens || tokens.length === 0) {
  logging.fatal(
    "{red There is no token to login to, please check your configuration.}"
  );
  logging.fatal("{red Quitting...}");
  process.exit();
}
if (useMain !== "true" && useMain !== "false") {
  logging.warning(
    "{rgb(255,245,107) useMain is not set correctly or is undefined. Defaulting to false.}"
  );
}
if (legitimacyCheck !== "true" && legitimacyCheck !== "false") {
  logging.warning(
    "{rgb(255,245,107) legitimacyCheck is not set correctly or is undefined. Defaulting to false.}"
  );
}
if (obfuscationCheck !== "true" && obfuscationCheck !== "false") {
  logging.warning(
    "{rgb(255,245,107) obfuscationCheck is not set correctly or is undefined. Defaulting to false.}"
  );
}
if (notesCheck !== "true" && notesCheck !== "false") {
  logging.warning(
    "{rgb(255,245,107) notesCheck is not set correctly or is undefined. Defaulting to false.}"
  );
}
if (permanentCache !== "true" && permanentCache !== "false") {
  logging.warning(
    "{rgb(255,245,107) permanentCache is not set correctly or is undefined. Defaulting to false.}"
  );
} else if (permanentCache === "true") {
  if (!fs.existsSync("./usedTokens.json")) {
    logging.info("{blueBright There is no usedTokens cache to load.}");
  } else {
    const file = fs.readFileSync("./usedTokens.json");

    if (file.length !== 0)
      try {
        usedTokens = JSON.parse(file);
      } catch (e) {
        logging.info(
          `{red Couldn't load usedTokens cache because of error: ${e}.}`
        );
        permanentCache = "false";
      }
    logging.info(
      `{blueBright Successfully read ${usedTokens.length} tokens from the usedTokens cache.}`
    );
  }
}

if (writeNotes !== "true" && writeNotes !== "false") {
  logging.warning(
    "{rgb(255,245,107) writeNotes is not set correctly or is undefined. Defaulting to false.}"
  );
} else if (writeNotes === "true") {
  if (!fs.existsSync("./notes")) fs.mkdirSync("./notes"); // Create notes folder if it doesn't exist
}

if (replit !== "true" && replit !== "false") {
  logging.warning(
    "{rgb(255,245,107) replit is not set correctly or is undefined. Defaulting to false.}"
  );
} else if (replit === "true") {
  logging.info("{blueBright Running in replit mode.}");
  // eslint-disable-next-line global-require
  const replit_http = require("http");

  replit_http
    .createServer((req, res) => {
      res.write("I'm alive!");
      res.end();
    })
    .listen(8080);
}
const paymentsourceid = (() => {
  const ressyncq = syncrq(
    "GET",
    "https://discord.com/api/v6/users/@me/billing/payment-sources",
    {
      headers: {
        authorization: mainToken,
        "user-agent": global.userAgent,
      },
    }
  );

  const ps = JSON.parse(ressyncq.body.toString());

  if (ps.message === "401: Unauthorized") {
    logging.fatal(`{red Main token not valid: ${ps.message}.}`);
    logging.fatal("{red Quitting...}");
    process.exit();
  } else if (ps.length === 0) {
    logging.warning(
      "{rgb(255,245,107) Main token does not have a billing source, some codes will not be sniped.}"
    );
    return "null";
  } else if (ps[0]) {
    logging.info("{blueBright Successfully got the payment method!.}");
    return ps[0].id;
  } else {
    logging.fatal(`{red Unable to get billing source: ${JSON.stringify(ps)}.}`);
    logging.fatal("{red Quitting...}");
    process.exit();
  }
})();

// eslint-disable-next-line no-restricted-syntax
for (const token of tokens) {
  const client = new Client({
    // https://discord.js.org/#/docs/main/v12/typedef/ClientOptions
    cacheGuilds: true,
    cacheChannels: false,
    cacheOverwrites: false,
    cacheRoles: false,
    cacheEmojis: false,
    cachePresences: false,
    messageCacheLifetime: 1,
    messageCacheMaxSize: 2,
    messageSweepInterval: 5,
    disabledEvents: [
      "GUILD_UPDATE",
      "GUILD_MEMBER_ADD",
      "GUILD_MEMBER_REMOVE",
      "GUILD_MEMBER_UPDATE",
      "GUILD_MEMBERS_CHUNK",
      "GUILD_ROLE_CREATE",
      "GUILD_ROLE_DELETE",
      "GUILD_ROLE_UPDATE",
      "GUILD_BAN_ADD",
      "GUILD_BAN_REMOVE",
      "CHANNEL_UPDATE",
      "CHANNEL_PINS_UPDATE",
      "MESSAGE_DELETE",
      "MESSAGE_UPDATE",
      "MESSAGE_DELETE_BULK",
      "MESSAGE_REACTION_ADD",
      "MESSAGE_REACTION_REMOVE",
      "MESSAGE_REACTION_REMOVE_ALL",
      "USER_UPDATE",
      "USER_NOTE_UPDATE",
      "USER_SETTINGS_UPDATE",
      "PRESENCE_UPDATE",
      "VOICE_STATE_UPDATE",
      "TYPING_START",
      "VOICE_SERVER_UPDATE",
      "RELATIONSHIP_ADD",
      "RELATIONSHIP_REMOVE",
    ],
  });

  client.on("message", async (msg) => {
    if (msg.author.id === client.user.id) return; // We don't want to snipe our own messages
    let codes = msg.content.match(global.regex);
    if (!codes || codes.length === 0) {
      codes = [];
    }
    if (msg.embeds.length > 0) {
      msg.embeds.forEach((embed) => {
        if (embed.fields) {
          // eslint-disable-next-line no-restricted-syntax
          for (const field of embed.fields) {
            codes.push(String(field.name).match(global.regex));
            codes.push(
              String(field.value)
                .replace(/[\])]$/gm, "")
                .match(global.regex)
            );
          }
        }
        if (embed.author) {
          if (embed.author.name) {
            codes.push(String(embed.author.name).match(global.regex));
          }
        }
        if (embed.description) {
          codes.push(String(embed.description).match(global.regex));
        }
        if (embed.footer) {
          if (embed.footer.text) {
            codes.push(String(embed.footer.text).match(global.regex));
          }
        }
        if (embed.title) {
          codes.push(String(embed.title).match(global.regex));
        }
      });
      codes = codes
        .filter((e) => e !== "null")
        .filter(Boolean)
        .flat();
    }
    if (!codes || codes.length === 0) {
      if (notesCheck === "false") return;
      const test = function () {
        switch (true) {
          case privnote.test(msg.content):
            return notes.privnote.getData(
              msg,
              writeNotes,
              msg.content.match(privnote).splice(0, 1).toString(),
              notes_webhookclient,
              client.user.tag
            );
          case temp_pm.test(msg.content):
            return notes.temp_pm.getData(
              msg,
              writeNotes,
              msg.content.match(temp_pm).splice(0, 1).toString(),
              notes_webhookclient,
              client.user.tag
            );
          default:
            return [];
        }
      };
      codes = codes.concat(await test()).filter((e) => e); // Concat and remove empty values
      if (!codes || codes.length === 0) return;
    }
    if (codes.length > 10) return;
    // eslint-disable-next-line no-restricted-syntax
    for (let code of codes) {
      const start = new Date();
      code = code.replace(
        /(discord\.gift\/|discord\.com\/gifts\/|discordapp\.com\/gifts\/)/gim,
        ""
      );
      if (obfuscationCheck === "true") {
        const code_no_symbols = code.replace(/\W/g, "");
        const code_no_obfuscation = code.replace(/\W.*/g, "");
        if (code_no_symbols !== code_no_obfuscation) {
          if (code_no_symbols.length > 26 && code_no_symbols.length < 16)
            code = code_no_symbols;
          else if (code_no_obfuscation.length > 26 && code_no_obfuscation < 16)
            code = code_no_obfuscation;
        } else if (code_no_symbols !== code) code = code_no_symbols;
      }
      if (legitimacyCheck === "true") {
        // const numeric = code.replace(/[^0-9]/g, "").length;
        const lowercase = code.replace(/[^a-z]+/g, "").length;
        const uppercase = code.replace(/[^A-Z]+/g, "").length;

        if (
          code.length > 26 ||
          code.length < 16 ||
          lowercase === 0 ||
          uppercase === 0 ||
          (lowercase < 4 && uppercase < 4)
        ) {
          // Very basic but no false positives. Will try to improve it in the future
          logging.success(
            `{rgb(137,96,142) Sniped [${code}] - Fake Code - ${
              msg.guild ? msg.guild.name : "DM"
            } from ${msg.author.tag}.}`
          );
          // eslint-disable-next-line no-continue
          continue;
        }
      }

      if (usedTokens.includes(code)) {
        logging.success(
          `{rgb(255,228,138) Sniped [${code}] - Already checked - Seen in ${
            msg.guild ? msg.guild.name : "DM"
          } from ${msg.author.tag}.}`
        );
        // eslint-disable-next-line no-continue
        continue;
      }
      http({
        url: `https://discord.com/api/v8/entitlements/gift-codes/${code}/redeem`,
        method: "POST",
        parse: "json",
        headers: {
          Authorization: mainToken,
          "User-Agent": global.userAgent,
          "Content-Type": "application/json",
        },
        data: {
          channel_id: msg.channel.id,
          payment_source_id: paymentsourceid,
        },
      })
        .then((res) => {
          const end = `${new Date() - start}ms`;
          if ("subscription_plan" in res.data) {
            logging.success(
              `{rgb(28,232,41) Sniped [${code}] - Success! - ${
                res.data.subscription_plan.name
              } - ${msg.guild ? msg.guild.name : "DM"} from ${
                msg.author.tag
              } - ${end}.}`
            );
            nitro_webhookclient.send_webhook_nitro(
              res.data.subscription_plan.name,
              msg.guild ? msg.guild.name : "DMs",
              msg.author.tag,
              client.user.tag,
              end,
              code,
              msg.url,
              webhookping_userid
            );
          } else {
            logging.error(
              `{red Tried to redeem code [${code}] but got error: ${res.data.message}.}`
            );
          }
        })
        .catch((err) => {
          const end = `${new Date() - start}ms`;
          if (err.response) {
            if (err.response.status === 401) {
              logging.error(
                `{red Tried to redeem code [${code}] but the main token is not valid.}`
              );
            } else if (
              err.response.data.message ===
                "This gift has been redeemed already." ||
              err.response.data.message === "Missing Access"
            ) {
              logging.success(
                `{rgb(255,228,138) Sniped [${code}] - Already redeemed - ${
                  msg.guild ? msg.guild.name : "DM"
                } from ${msg.author.tag} - ${end}.}`
              );
            } else if (
              err.response.data.message ===
              "You need to verify your e-mail in order to perform this action."
            ) {
              logging.error(
                `{red Tried to redeem code [${code}] but the main token doesn't have a verified e-mail.}`
              );
            } else if (err.response.data.message === "Unknown Gift Code") {
              logging.success(
                `{redBright Sniped [${code}] - Invalid - ${
                  msg.guild ? msg.guild.name : "DM"
                } from ${msg.author.tag} - ${end}.}`
              );
            } else if (
              err.response.data.message ===
                "New subscription required to redeem gift." ||
              err.response.data.message === "Already purchased"
            ) {
              logging.error(
                `{red Tried to redeem code [${code}] but the gift type cannot be used with an existing Nitro.}`
              );
            } else if (
              err.response.data.message ===
              "Payment source required to redeem gift."
            ) {
              logging.success(
                `{rgb(28,232,41) Sniped [${code}] - You don't have a valid payment method.}`
              );
            }
          } else {
            logging.error(
              `{red Tried to redeem code [${code}] but got connection error: ${err}}`
            );
          }
        });
      usedTokens.push(code);
      if (permanentCache === "true")
        fs.writeFileSync("./usedTokens.json", JSON.stringify(usedTokens));
    }
  });

  client.on("ready", () => {
    if (token === mainToken) {
      logging.info(
        `{cyan (INFO)} {blue Main token valid: ${client.user.tag} - Sniping in ${client.guilds.cache.size} servers.}`
      );
    } else {
      logging.info(
        `{cyan Slave logged in as ${client.user.tag} - Sniping in ${client.guilds.cache.size} servers.}`
      );
    }
    if (token !== mainToken) client.user.setStatus(tokenStatus);
    else client.user.setStatus(mainStatus);
  });

  client.on("shardError", (error) => {
    logging.error(
      `{red Token [${token.substring(
        0,
        10
      )}...] encountered a connection error: ${error}.}`
    );
    client.on("shardResume", () => {
      logging.info(
        `{blueBright Token [${token.substring(
          0,
          10
        )}...] reconnected successfully!}`
      );
    });
  });

  setTimeout(() => {
    client.login(token).catch((err) => {
      if (token === mainToken) {
        logging.fatal(`{red Main token not valid: ${err}.}`);
        logging.fatal("{red Quitting...}");
        process.exit();
      } else {
        logging.error(
          `{red Skipping slave token "${token.substring(0, 10)}...": ${err}.}`
        );
        clearTimeout();
      }
    });
  }, Math.floor(Math.random() * (1100 * tokens.length - 500 + 1)) + 500);
}
